src/python_testing/TC_IDM_2_2.py

```diff
@@ -0,0 +1,669 @@
+                                        msg="Unexpected success reading invalid attribute")
+                    logging.info(f"Confirmed unsupported attribute {unsupported_attr} returns error on endpoint {endpoint_id}")
+                    return
+
+        # If we get here, we got problems as there should always be at least one unsupported attribute
+        asserts.fail("No unsupported attributes found we must find at least one unsupported attribute")
+
+    async def _read_repeat_attribute(self, endpoint, cluster, attribute, repeat_count):
+        results = []
+        for i in range(repeat_count):
+            path = AttributePath(EndpointId=endpoint, ClusterId=cluster.id, AttributeId=attribute.attribute_id)
+            result = await self.verify_attribute_read([path])
+            results.append(result)
+
+        # Verify all reads returned consistent values
+        if len(results) > 1:
+            first_result = results[0]
+            for i, result in enumerate(results[1:], 2):
+                # Compare the attribute values from each read
+                first_attr_value = first_result.tlvAttributes[endpoint][cluster.id][attribute.attribute_id]
+                current_attr_value = result.tlvAttributes[endpoint][cluster.id][attribute.attribute_id]
+                asserts.assert_equal(first_attr_value, current_attr_value,
+                                     f"Read {i} returned different value than first read")
+
+        logging.info(f"Successfully completed {repeat_count} consistent reads of {attribute}")
+        return results
+
+    async def _read_data_version_filter(self, endpoint, cluster, attribute, test_value=None):
+        read_request = await self.default_controller.ReadAttribute(
+            self.dut_node_id, [(endpoint, cluster, attribute)])
+        data_version = read_request[0][cluster][Clusters.Attribute.DataVersion]
+        if test_value is not None:
+            await self.default_controller.WriteAttribute(
+                self.dut_node_id,
+                [(endpoint, attribute(value=test_value))])
+        data_version_filter = [(endpoint, cluster, data_version)]
+        filtered_read = await self.default_controller.ReadAttribute(
+            self.dut_node_id,
+            [(endpoint, cluster, attribute)],
+            dataVersionFilters=data_version_filter)
+        return read_request, filtered_read
+
+    async def _read_non_global_attribute_across_all_clusters(self, endpoint=None, attribute=None):
+        attribute_path = AttributePath(
+            EndpointId=endpoint,
+            ClusterId=None,
+            AttributeId=attribute.attribute_id)
+        try:
+            await self.default_controller.ReadAttribute(
+                self.dut_node_id,
+                [attribute_path]
+            )
+            asserts.fail("Expected INVALID_ACTION error but operation succeeded")
+        except ChipStackError as e:  # chipstack-ok
+            asserts.assert_equal(e.err, 0x580,

```

Comment:
```suggestion
            asserts.assert_equal(e.err, 0x80,
```

Could you double check this code? According to the spec here (https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/master/src/data_model/Interaction-Model.adoc#101-status-code-table) INVALID_ACTION is `0x80` instead of `0x580`

Reply:
Oh, I see, thanks! Maybe it would be a good idea to actually list all these error codes in a file with named constants/enum that can be reused? That way, people reading the code, don't have to check that the code is correct, they will see something like `ERROR_CODES.INVALID_ACTION`, etc.

Reply:
This can be done as a follow up, though, so don't need to do this now. Just a suggestion that might help all of the test implementations.